### ISSUE TITLE
Linux / OS X Support (SteamVR & Desktop)

### ISSUE THREAD

**Issue by Frooxius** [Neos-Metaverse/NeosPublic#149]
Make sure that Neos builds run on Linux and/or OS X and put them into the compilation pipeline, so they're updated together with the Windows build.

This will require some extra work to setup, but should be issue free once it's setup. However I need to gauge the interest in this feature and prioritize accordingly.

How many people here would need/want this?

**Comment by Casuallynoted** [comment]
I'd definitely be interested in this- not super super high priority, but I've been looking at the possibility of slowly switching all my operations to Linux once I've moved my VR setup over to SteamVR (Valve Index, woot!) as Rift doesn't have compatibility with anything outside of PC. Tbh even Proton support in the time being (idk if that can be done on the Dev's side though, may be a Valve thing) would be super useful :D 

In the meantime though, there are a few other VR titles on steam that would need to get decent proton support/linux builds before I made that switch so not super high priority :)

**Comment by yellowcrash10** [comment]
I'm very interested in Linux support. All of my machines run Linux, so it would be the only way I could run the program.

It looks like it's close to being able to run with Proton, so that could be a good workaround in the meantime until a proper Linux build is done.

**Comment by codecnotsupported** [comment]
> I'm very interested in Linux support. All of my machines run Linux, so it would be the only way I could run the program.
> 
> It looks like it's close to being able to run with Proton, so that could be a good workaround in the meantime until a proper Linux build is done.

For me it crashes on startup with Proton. With the latest Wine version it is semi stable.
That aside, I'm as well interested in native Linux support.

**Comment by inzanesonikku** [comment]
now since i've got a server that runs lubuntu, i am VERY much interested in having a linux version :>

i wanna try to host a world or two on it

**Comment by AdmiralGhoster** [comment]
https://github.com/Frooxius/NeosPublic/issues/149#issuecomment-514208650

If you want to host a world, i would suggest looking at the headless client issue/feature and commenting how it might be helpful, as right now you just open a second client (generally offsite) and run ot there but with the headless and linux support you could host bigger world with more users (since networking is peer-to-peer)

**Comment by inzanesonikku** [comment]
oh ye you right. sorta forgot about the headless client thing

**Comment by fabiankranewitter** [comment]
I would be happy if there is a linux version.

**Comment by Frooxius** [comment]
Thanks for the feedback! Are you guys mainly looking into Linux version for the headless client, or to run it with VR headset?

**Comment by STrRedWolf** [comment]
VR Headset. My main rig is Linux. I even run Second Life desktop client there.

**Comment by BeckersC** [comment]
I would like to run the full experience on linux eventually but with the current state of support for VR devices being lacking (Only Vive and index) you really shouldn't spend too much time on us.

**Comment by fabiankranewitter** [comment]
Vr headset support would be nice.

**Comment by codecnotsupported** [comment]
I'd like the full Vr experience as well.

**Comment by yellowcrash10** [comment]
I would like to run it with a VR headset.

**Comment by Frooxius** [comment]
I see, thanks for the feedback. I think full VR client is doable (although I need to investigate more in the Linux VR support with SteamVR, I'm not sure what the state is). 

However one major blocker is OVRLipSync library, which is used for viseme analysis, as it doesn't support Linux. I'd have to either find an alternative that works under Linux.

Alternatively it could just be randomly moving modulated by the volume, so it wouldn't be accurate at all.

The other native libraries that Neos is using should compile for Linux and work there, so any blocker would be any extra compatibility issues that might pop up, at least from what I can think of now.

If anyone knows a good alternative to OVRLipSync that could be used please let me know.

**Comment by Casuallynoted** [comment]
I’m honestly not immediately familiar with any myself, I found this asset but I’m not sure how good it would be for this specific use case-

https://assetstore.unity.com/packages/tools/animation/salsa-lipsync-suite-148442

**Comment by Frooxius** [comment]
I was actually looking at that one, but I'm not sure how good fit it is. It seems to have a lot of extra stuff and I'm not sure if I can use it as standalone and just pipe my raw audio data into it and get analyzed visemes out. Anyone have experience with it?

**Comment by fabiankranewitter** [comment]
Have another software found here
, There is also a free lite version and it is cross platform compatible. https://lipsync.rogodigital.com/

**Comment by Casuallynoted** [comment]
I really love the idea of the Rogodigital one, but atm it sadly doesn’t support real-time sync which would be a necessity for Neos.

**Comment by codecnotsupported** [comment]
> I see, thanks for the feedback. I think full VR client is doable (although I need to investigate more in the Linux VR support with SteamVR, I'm not sure what the state is).
> 
> However one major blocker is OVRLipSync library, which is used for viseme analysis, as it doesn't support Linux. I'd have to either find an alternative that works under Linux.
> 
> Alternatively it could just be randomly moving modulated by the volume, so it wouldn't be accurate at all.
> 
> The other native libraries that Neos is using should compile for Linux and work there, so any blocker would be any extra compatibility issues that might pop up, at least from what I can think of now.
> 
> If anyone knows a good alternative to OVRLipSync that could be used please let me know.

The documentation [here](https://developer.oculus.com/documentation/audiosdk/latest/concepts/audio-ovrlipsync-sample/) says it has a Linux build option for Oculus Lipsync.

**Comment by NerveCoordinator** [comment]
I would also love a linux version for this!

**Comment by aaronfranke** [comment]
Hey, I'm interested in Linux support for Neos VR. I mainly run Linux so it would be awesome to run Neos VR on my Linux system. If desired, I can help with porting and/or testing a port, I have experience with Linux, Unity, and C#, though I have no idea about lip sync libraries. 

By the way, I'm not sure if you were aware, but you can leave reactions on GitHub comments. I've added a thumbs up to the top post. This can help keep count of the number of interested people.

**Comment by Frooxius** [comment]
Thanks for the info!  The lip sync library right now is the biggest issue that needs to be resolved. The rest of the native libraries that Neos is using I can already compile from source for Linux.

Rest of Neos should generally work, save for some platform specific hurdles, but I don't think these will be too much to deal with.

**Comment by Frooxius** [comment]
Did some preliminary work on a Linux build. Compiled Freeimage and Opus for Linux (x64 arch) and fixed up a few issues (e.g. Leap Motion won't be supported since they don't have Linux).

The build partially boots, but currently fails due to the shaders not being compiled for Linux. I'll try to set them up soon, but they take several hours to all compile. 

I've created a new channel on the [Discord](https://discord.gg/NeosVR) server #linux-preview If you're a Patreon supporter you'll have access to it right away, if not let me know and I can give you access.

You can get the build there, but it's not usable yet, but you can at least see if it boots up on you to at least something (some text should appear).

**Comment by LatePacket** [comment]
Awesome glad to hear it's coming along, will have to try it out soon 😃

**Comment by Frooxius** [comment]
> > The build partially boots, but currently fails due to the shaders not being compiled for Linux. I'll try to set them up soon, but they take several hours to all compile.
> 
> If you aren't already, I suggest using Vulkan, both on Windows and Linux. Vulkan has better performance than OpenGL and DX11, and then you don't have to compile some shaders twice.

I'd like to update to it in the future, but that's not possible at the moment. The Vulkan implementation in Unity isn't mature enough for that and there are come compatibility issues - the libVLC based video player library doesn't support it for example.

**Comment by Frooxius** [comment]
The latest version of the build with a bunch of things disabled (VLC video playback, Assimp mainly) and the shaders compiled seems to boot and run. There are some weird issues (dash is not clickable) and it crashes when trying to connect to a session.

It's some progress, but it seems it requires a lot more work than I was hoping for, so it will take longer to get an usable functioning build.

Also this is a bit random and I don't plan on working on OS X build anytime soon (unless there's also demand for it), but I was digging through things and found this really old alpha build in case anyone wants to play with it:
https://cloudxstorage.blob.core.windows.net/install/NeosAlphaMac.zip

**Comment by Casuallynoted** [comment]
I am definitely playing with that :P Thank you so much for all your hard work Frooxius!!

**Comment by BeckersC** [comment]
I experimented a bit with the current windows branch running through proton 4.11-3 on linux (intel + amdgpu pro) running in screen mode. (Oculus still doesnt like linux)

To my surprise mostly everything worked perfectly fine. I could import 3d Models, switch world, watch videos, grab objects, switch avatars and interact with the dash.

That is until I installed steamvr which makes the dash  unresponsive. That was the only issue that I could find.

**Comment by Frooxius** [comment]
I managed to fix the dash not being responsive, it turned out to be a subtle difference in handling of floating point numbers, which broke approximate comparisons (so even using approximate comparison of zero against zero would be false). It's going to be in the next build, should help with Proton.

I have also found out what's causing the freezes/crashes of the native builds. Unfortunately it's a severe bug in Unity itself, particularly their audio thread which Neos hooks to do all its audio work.

I have filed a bug report with Unity: https://fogbugz.unity3d.com/default.asp?1181376_n4n19nnl5bgl2t1v Unfortunately until this is fixed on their end any work on Linux build is blocked.

**Comment by aaronfranke** [comment]
When it comes to floating-point number comparison, here's a simple piece of code for it:

        public static bool IsEqualApprox(real_t a, real_t b)
        {
            // Check for exact equality first, required to handle "infinity" values.
            if (a == b) {
                return true;
            }
            // Then check for approximate equality.
            real_t tolerance = Epsilon * Abs(a);
            if (tolerance < Epsilon) {
                tolerance = Epsilon;
            }
            return Abs(a - b) < tolerance;
        }

**Comment by Frooxius** [comment]
Thanks, but I'm not looking for piece of code, like I said it's already fixed.

**Comment by Frooxius** [comment]
I've got some good news, Unity has managed to reproduce the issue! 
https://issuetracker.unity3d.com/issues/linux-player-stack-overflow-crash-when-accessing-audiosettings-dot-dsptime-in-onaudiofilterread

If you have Unity account you can vote on it to give it a higher priority.

**Comment by Frooxius** [comment]
I have some more good news! The headless client/server version of Neos is now available for Linux! https://github.com/Frooxius/NeosPublic/issues/178

You can install it on Steam in a beta branch. Currently it's in closed beta. Join our Discord to participate, Patreons have automatic access, others based on invite.

**Comment by jcc10** [comment]
> **Unfortunately until this is fixed on their end any work on Linux build is currently blocked.**

Supposedly fixed in 2020.1

**Comment by Casuallynoted** [comment]
@jcc10 We had some discussion about this in the Linux build channel, and it’s kind of an odd situation at the moment. None of the 2020 alpha logs show the fix being added, so it may be an internal build at the moment or poor documentation. They also haven’t backported it to more stable versions. I know Froox said the most experimental he’d want to go is a beta build like 2019.3. I reached out to Unity on the issue page to try to get some clarificationb

**Comment by source83** [comment]
Only JUST bumped into NeosVR today and am very intrigued by it's capability's.
I am also a heavy linux user and finding it very disappointing (not meant toward NeosVR or Froox) about how little attention there is toward us linux users.
Don't get me wrong, I UNDERSTAND that it's because the overwhelming majority of games/programs are made for windows, but still...
Not to mention Oculus dropping Linux support completely in the pre CV1/DK2 stages... THAT really got my gears grinding :S

**Comment by Casuallynoted** [comment]
> Only JUST bumped into NeosVR today and am very intrigued by it's capability's.
> I am also a heavy linux user and finding it very disappointing (not meant toward NeosVR or Froox) about how little attention there is toward us linux users.
> Don't get me wrong, I UNDERSTAND that it's because the overwhelming majority of games/programs are made for windows, but still...
> Not to mention Oculus dropping Linux support completely in the pre CV1/DK2 stages... THAT really got my gears grinding :S

No I totally agree with this. Neos has been a massive breath of fresh air in this department because most developers are like "Not worth it" the second Linux is brought up. There are definitely considerations to be made and I don't want to see Linux prioritized over absolutely essential features for any reason, but the fact that it's something Froox has made major efforts towards in spite of Unity and Oculus' shortcomings is a big big deal to me. Froox is awesome. :)

**Comment by jcc10** [comment]
What is disappointing is that the unity bug report list's it as solved but not a single changelog mentions the bug. I made a comment on the bug report to that fact.

I just want to confirm that it wasn't silently fixed in one of the beta's for 2020.1 and there just hasn't been any tests. I don't do unity development so I don't know if that is a doable test to do.

If it is still in the alpha over a month after they flagged it as fixed the only solution may be to submit a new issue referencing the old one never getting fixed. (Possibly point them over here as well to show that there are people literally waiting on the fix.)

As a side note: I was able to boot NEOS in photon 4.2-9, it loaded to the local lobby but the headset model was on the ground. I'm going to do more testing tomorrow or over the week end.

Hopefully we will have a partial version until Unity decides to actually explain where the bugfix is.

**Comment by Casuallynoted** [comment]
>What is disappointing is that the unity bug report list's it as solved but not a single changelog mentions the bug. I made a comment on the bug report to that fact. ...

@jcc10 I wanna say that if they haven’t added the fix to any documentation at this point they probably haven’t added the fix to unity either, I would second the idea of an additional issue on Unity’s bug tracker explaining that they’ve never actually documented or released a fix.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by source83** [comment]
I'm guessing there has been no progress since Unity still has not addressed the issue itself.
Btw, is it too late or inappropriate to ask for access to the #linux-preview channel in discord ?

**Comment by Frooxius** [comment]
Sorry no news. I'm also currently busy with UI overhaul, so I don't have much time to look into Linux support yet. I can give you access, let me know your username on Discord.

**Comment by source83** [comment]
macromarco

**Comment by Robertus-Rob** [comment]
Vote for Desktop mode on OS X, although I understand not much users for them at this moment

**Comment by Frooxius** [comment]
I did a new build with latest version of Unity, it seems the blocking issue was fixed and the sound works again now!

There are some things that still need to be done now. Mainly the asset variant system needs to be ported to Linux right now so the build can function. I'm not sure when I'll get the time to work on that, but at least the major showstopper seems to be gone now.

**Comment by Frooxius** [comment]
I've got some good news, Linux build is working!

https://www.youtube.com/watch?v=1AEZ17C5TQU

It's available on Steam on the default branch (install on Linux to get the native Linux build). 

It still needs quite a bit of work to get it on/near parity with the Windows version. Currently following doesn't work:

-- Most of video playback (any media services like YouTube don't work, as well as any formats played by the libVLC)
-- 3D model importing is not supported at the moment
-- Copy & Paste doesn't work - you need to use the built-in file browser to imprort assets
-- Visemes don't work. This is unlikely to get fixed soon due to the library not supporting Linux, might add jaw flap simulation
-- Flat Lit Toon shader doesn't work. If anyone would like to have a stab at getting it to work let me know, that could save me some time
-- SteamVR is untested

**Comment by Casuallynoted** [comment]
I can test it with SteamVR in a bit, one thing I do know is that Flat Lit Toon has been deprecated by the original creator- they’ve got a doc with some more up to date replacements if that helps!

**Comment by Frooxius** [comment]
Updated Linux build (as well as all others), latest build is 2020.3.12.771

- Microphone now works
- Copy & Paste works (3D model import still won't work)

I've looked into video playback, but UMP support on Linux is horribly broken. Might be better to wait for official libVLC to support Linux with Unity. If it's crucial for you, I can contact UMP support and see if they can help.

**Comment by jug007** [comment]
So I have just run the game once:
Text does not show. I guess it is using a font that is not installed on my system and doesn't have proper fallback.
Also there is some weird shading problems on the world I spawned in. Some of the cubed look seethrough and the menu has like a reflection of itself in it which moves around weirdly.
There is also no logs in /home/user/.local/share/ as far as I can tell. I don't know where they are supposed to be though.

**Comment by BeckersC** [comment]
Using Desktop mode: No unexpected weird glitches. Everything appears to be working optimally. No trasparent textures. Local videoplayback and sound is working. 

The FileBrwoser tool works too although some of the shortcuts to different harddrives and mountpoints are broken and show as empty. Navigating to them manually works. Not a big issue.

It would be awesome if remote video playback and 3d imports would be worked on in the future :)
As already advertised .obj imports as a black silhouette of the 3d model which I believe can't be interacted with while .fbx and .blend fail during decode.

System info: PopOS 20.04 beta, intel, nvidia proprietary driver, lots of ram

**Comment by Enverex** [comment]
@BeckersC Can you go to "The Cabin Lair" (there should be a public session open) and make sure you can actually see the ground? Everyone else is reporting that the PBS Splat Shader doesn't work, but it's not used in many places.

**Comment by BeckersC** [comment]
@Enverex I confirm the shader used on the forest ground in the Cabin Lair appears to be broken on linux.

**Comment by Frooxius** [comment]
@jug007 What system are you running on? Any change to get logs?

Neos doesn't use system fonts at all, it downloads its own TTF files, so it's definitely not that. It's likely some of the native libraries (Freetype.so, msdfgen.so) used by the text rendering not working properly. A log would likely indicate if that's the case, you can find it in ~/.config/unity3d/Solirax/NeosVR/Player.log

If you're experiencing shading issues with Flat Lit Toon shader, that's a known issue, it doesn't work under Linux. See my original post for other known issues.

@BeckersC That's interesting that video playback is working for you. What kind of videos have you loaded? What exactly do you mean by remote playback? Things like YouTube URL's? Or RTMP streams and such?

3D import is currently broken now, it's mostly issue with the native dependency on Assimp. It'll be worked on at some point in the future once it gets enough priority.

Interesting that the PBS Splat Shader still doesn't work, it was simply missing the shader, but it should be there now. Which build did you test with?

**Comment by jug007** [comment]
```
Plugins: Couldn't open /media/motofckr9k/EffolenaMKII/SteamLibrary/steamapps/common/NeosVR/Neos_Data/Plugins/libFreeimage.so, error: /media/motofckr9k/EffolenaMKII/SteamLibrary/steamapps/common/NeosVR/Neos_Data/Plugins/libFreeimage.so: cannot open shared object file: No such file or directory 
(Filename:  Line: 306
```
This seems like it *might* be the problem then.
Full log here: [Player.log](https://github.com/Frooxius/NeosPublic/files/4343785/Player.log)

**Comment by Frooxius** [comment]
Hmm that's pretty bizarre. It seems to be requesting libFreeimage.so instead of libFreeImage.so for some reason (which is the actual library). I haven't seen that behavior before.

Any idea what could be causing that? libFreeImage.so is the one it requests from code and loads on my testing system, it almost seems like there's something intercepting the request and changing the case on your system.

You could try renaming the library in the Plugins folder to see if it loads.

**Comment by jug007** [comment]
It does say that it `Preloaded 'libFreeImage.so'` in that logfile.

When renaming the file to `libFreeimage.so` the game crashes on start.
[Player.log](https://github.com/Frooxius/NeosPublic/files/4343854/Player.log)
I really don't think that there is any way that "something intercepting the request" could be a thing.
There is only two things I can think of, when it comes to it not working on my system, but it working on yours: either your system doesn't care about case sensitivity (which would be weird for a linux system), or there is some really weird problem with the steam runtime. There is basically two steam runtimes right now; the one installed when you install steam via the .deb package from valve directly and a slightly modified one that is in the ubuntu package repositorys.
I guess also there is the difference that you said you didn't try SteamVR and I have only tried it in VR for now.
Currently I am having problems with starting NeosVr in desktop mode, but that might just be steam being janky right now.

**Comment by Frooxius** [comment]
VR definitely should be working, it's been tested by some users. I don't think it's Steam either, since it's Neos itself that uses the libFreeImage.so library. There's a fairly complex logic in dynamically loading libraries both in Mono (but that should be the same since it's part of the build) and possibly the system. 

Mono itself definitely tries a lot of different names and variations when it can't find the library. Could be some other problem where the library itself cannot be loaded for some reason and that's just what it fails on.

What if you make two copies of libFreeImage.so? One with the lowercase and one with uppercase?

**Comment by jug007** [comment]
When I make two copies it preloads both: [Player.log](https://github.com/Frooxius/NeosPublic/files/4343954/Player.log)
Text shows now. 
Is there a screenshot button? The text is really hard to read because of it being small and the shading making the backgroud behind the black text very dark.

**Comment by Frooxius** [comment]
You can use the camera from the dash. Does regular PrintScreen not work though? It's odd with the shading though. I'd definitely need to see what it looks like, sounds like there are other things that are wrong.

**Comment by jug007** [comment]
Yeah I forgot I could have just used the screenshot utility of the OS since it shows on the desktop as well.
Here is a video of said behaviour: http://motofckr9k.ddns.net/m/neos.mkv

**Comment by BeckersC** [comment]
@Frooxius  Unless steam has updated neos since my test a couple of hours ago I was on the beta 2020.3.15.1036

What I call local videoplayback is a videofile imported through the filebrowser tooltip. I retested now with multiple different video files and formats from mp4, mov to mkv created from varying applications like blender, obs, ffmpeg and unknown internet sources. Playback, video scrubbing, audio and import worked flawlessly on everyone of them.

What I call remote video was the twitch and mixer video stream players from /Neos Essentials/VideoPlayers. I was not able to make them do anything except enter a url to a twitch/mixer channel.

**Comment by jug007** [comment]
There is also an issue with performance. I don't know if it has anything to do with the warning that steam pops up saying that NeosVR is running in desktop *and* vr mode, but performance is very bad considering the world I spawn in being pretty much devoid of anything other than flying cubes and my specs. Games which are known to be underperforming like VRChat for example run a lot better when it comes to performance. And that is without them even having a native version.

**Comment by Frooxius** [comment]
@jug007 Hmm on the video everything looks the way it's supposed to, I don't see any visual artifacts. The darker background is due to the lighting, if you rotate around or go to another world it should be fine. The upcoming UI overhaul will improve the look of the UI.

What hardware are you running on? Mainly your GPU and CPU? Are you using latest official drivers? It could be some weird interaction with the system given the other strange issues you've experienced. It'd be good if you could compare with the Windows version on the same machine. 

Neos itself doesn't do much differently on Linux on itself, so whatever performance problems are likely due to the OpenGL vs DirectX 11 implementation, the system, drivers and SteamVR itself.

Note that capping to 60 FPS is likely coming from SteamVR, it limits the framerate to half if some frames are dropped.

@BeckersC Ah. Twitch/Mixer and such require youtube-dl to work correctly. That's currently unsupported, I need to investigate if it's possible to include this with the build or require the users to install it on their system.

However video playback isn't technically supported yet either, I can't get it to work on my test system, so it's likely going to take a while.

Can you check which playback engine is used in the inspector though? Is it u sing Unity Native or libVLC?

**Comment by BeckersC** [comment]
@Frooxius The videoplayer is using libVLC by default on my end. When I switch over to Unity Native it stops working.

I have youtube-dl installed on my system but I guess it will need to be hooked up. 

There is no hurry in getting this to work on my end.

**Comment by jug007** [comment]
So the game is underperforming on windows as well, but it runs a lot better there.
I am hitting the 72fps (half) cap when in the traning center. Opening the camera makes it go down to 70.
On linux I have tried the same scenario. ~41fps with the camera open in the training center.
I am running an i7-7800x overclocked to 4,9GHz with 32GB of ram and a GTX 1080.
Driver wise I am using 440.64 which is the newest for my system. I forgot to check on windows, but I had just installed an update from nvidia a week ago.

For comparison: VRChat would hit the 144fps cap on windows and would hit 144fps without the camera on linux in a similarly empty room and 72fps with something like a camera or a mirror enabled.

But I guess windows performance isn't really an issue here.

**Comment by Frooxius** [comment]
@jug007 That sounds like it's running fine on Windows. It's a different rendering pipeline, postprocessing and environment, so you can't make one to one comparison. Hitting 144 FPS with GTX 1080 is very tricky and if it isn't maintained it'll be dropped to half, even if it could run faster than that. Getting frame times could help get more accurate data.

However the difference between the platforms is most important. It's strange that you're getting such performance drop on Linux, it could be for a lot of reasons. We need to gather more data from more setups and distro's, because I had conflicting reports of some people getting better performance on Linux vs Windows.

@BeckersC Huh interesting. On my testing system the libVLC will fail to load. It's good to know that it loads on some systems at least though, it might be that my Linux install is missing some dependencies.

**Comment by jcc10** [comment]
Right now Neos does not recognize the headset so it starts in pancake mode (With only debug textures)
https://youtu.be/vtEnByCA7hk?t=248

Pop!_OS (Ubuntu Derivative.)

Thanks for even working on this at all, it's more then most developers do.

**Comment by shiftyscales** [comment]
Hmm. I assume other apps are recognizing your hardware just fine? Just to check a couple of things- are you launching from the desktop, or from within SteamVR? You don't have any proton compatibility tools enabled so you will be running the native build?

You could try forcing the SteamVR runtime by using the `-SteamVR` command line argument. http://wiki.neosvr.com/subdom/wiki/index.php?title=Command_Line_Arguments#Forcing_specific_hardware

I'm still waiting a few weeks on my Index, and then I'll be able to hopefully provide some more in-depth help, and support for troubleshooting, and testing on the Linux builds. Let me know how things go, @jcc10.

If you aren't already on it- we have a Discord server with plenty of users that provide support. https://discord.gg/NeosVR

You can also reach out to me personally: @Shifty#7944

**Comment by jcc10** [comment]
Ok, issue was I had "Steam Linux Runtime" Selected. I guess I thought that meant "Force Native" for whatever reason?

Well It's working now. (Also scratched out last message.) Thanks @shiftyscales !

**Comment by guysoft** [comment]
@jug007 @Frooxius Just got SteamVR working on linux from Steam, using an HTC Vive. I have dual boot with windows and there is a performance drop. When I move the headset there is a considerable lag.
Is there any kind of debug output that would help you understand what might be the issue?

**Comment by shiftyscales** [comment]
I believe I'm aware of the issue, @guysoft. Are you using an Nvidia GPU by chance? In short- they haven't enabled features in their Linux driver that enable asynchronous reprojection to work properly in SteamVR. As such, unless you can maintain your full native refresh rate on the headset, there will be a lot of stuttering in the rendering.

https://twitter.com/dan_ginsburg/status/1261403868279140353
https://github.com/ValveSoftware/SteamVR-for-Linux/issues/214#issuecomment-629634692

**Comment by guysoft** [comment]
@shiftyscales Yes I am using Nvidia, I have a Asus Zephyrus g14 with an NVIDIA RTX 2060 Max-Q.
It also has an AMD on the integrated CPU. But its not strong enough to push VR.
Since this is a laptop I can't really switch cards neither.

Is there anything I can do to help?

**Comment by shiftyscales** [comment]
Not in this case, unfortunately @guysoft. It's a limitation of the Nvidia driver. So anyone using Nvidia on Linux with VR will need to wait for them to hopefully update their driver to support asynchronous reprojection.

For now, you are probably best playing Neos on Windows native. There are currently other limitations on the Linux branch as well due to some third party libraries being incompatible.

**Comment by TiagoTiago** [comment]
Would it be possible to have some sort of dynamic quality setting to ensure the framerate target is constantly reached, like automatic LODing with distance, reducing animation updates the further away something is from the camera and from the center of the field of view, switching to monoscopic rendering past a certain distance like how Unreal does (maybe it's only for mobile VR, I'm not too clear on that), low resolution and framerate billboard impostors after certain distance etc, lowering the whole rendering resolution with distance; and have the threshold/intensity adjusted automatically to ensure the framerate?

What about having separate world and GUI resolutions, so text is still legible even if you have huge pixels for the world?

**Comment by Frooxius** [comment]
@TiagoTiago There are variety of performance improvements and features planned, you can see some of them by searching the Performance tag here on GitHub: https://github.com/Frooxius/NeosPublic/issues?q=is%3Aissue+is%3Aopen+performance+label%3APerformance

This topic is focused on Linux / OS X support, so it's a bit out of scope. However I should note that in general that it's not possible to ensure that framerate target is always reached, since bottlenecks come from many different sources that are not possible to control in such way. Currently that's mostly CPU usage rather than GPU, so that's where the main focus is.

**Comment by TiagoTiago** [comment]
Should I post each of those ideas as separate issues?

**Comment by Frooxius** [comment]
We already have existing issues for most of these, so it's not really necessary. We're already focusing on things that help improve performance so other performance features will come naturally with that as well, particularly when they're common techniques.

**Comment by yellowcrash10** [comment]
I tried Neos in VR on Linux and ran into a peculiar issue. The left eye on the headset displayed the correct image, but the right eye displayed colored auras around certain objects. I tried to take a screenshot, but SteamVR wouldn't take one and I had to use my phone. Apologies for the bad quality.

Left eye:
![NeosEyeLeft](https://user-images.githubusercontent.com/1989498/104298269-61067d00-5489-11eb-8f82-1d920876aa31.jpeg)
Right eye:
![NeosEyeRight](https://user-images.githubusercontent.com/1989498/104298983-3ff25c00-548a-11eb-8c13-6aaef86a844e.jpeg)

**Comment by Fleur24** [comment]
Just an idea, regarding OS X (macOS), Apple M1 version in the future?

**Comment by Frooxius** [comment]
Possibly at some point, but right now even x86-64 version isn't too likely to be prioritized for OS X soon due to Apple dropping OpenGL and pretty much requiring Metal, which significantly complicates things on our end and adds a lot of workload. For M1 we'd require Unity to support it first as well (which is no their roadmap as far as I know), but things might change between then.

**Comment by yellowcrash10** [comment]
@BeckersC At one point you had video players working on the Linux build. Can you confirm if they still work? If they do work, could you describe your setup so that others might be able to get them working?

**Comment by BeckersC** [comment]
@yellowcrash10 I am sorry to report but video playback is not working for me currently. Bear in mind I have since switched to another distribution.

**Comment by guysoft** [comment]
I managed to start neos on Linux with an original vive. But it keeps crashing,  not sure why. Sometimes it manages to pass the splash screen. Not sure why it only works sometimes.

**Comment by Frooxius** [comment]
I recently found this library for LipSync. This should work on Linux, which would somewhat solve the missing Oculus Lip Sync, however it seems to be more limited, as it only detects some basic phonemes, rather than the full set.
https://github.com/hecomi/uLipSync

I haven't looked into it in detail yet though, just logging it here. I can look into integrating it when there's a chance, unless something else turns out.

Also I'm considering closing this megathread at this point. We have had Linux support for a while, just missing some things and it might be better to instead open those as individual issues now instead to better track them and make a separate one for Mac OS X, since that's a whole separate topic.

**Comment by shadowpanther** [comment]
> Also I'm considering closing this megathread at this point. We have had Linux support for a while, just missing some things and it might be better to instead open those as individual issues now instead to better track them and make a separate one for Mac OS X, since that's a whole separate topic.

Maybe make a separate roadmap (project) and pin the remaining Linux issues there to keep track in a more efficient/transparent way?

**Comment by Frooxius** [comment]
Yeah that sounds good thanks! I'm closing this one then and going to setup stuff here: https://github.com/Neos-Metaverse/NeosPublic/projects/20
